src/ahp/mod.rs

```diff
@@ -359,10 +356,149 @@ impl<F: PrimeField> UnnormalizedBivariateLagrangePoly<F> for Box<dyn EvaluationD
+        }
+
+        Ok( Self { poly: boundary_poly, evals: poly_evals } )
+    }
+
+    /// Return the underlying boundary polynomial, consuming `self`.
+    pub fn polynomial(self) -> DensePolynomial<F> { self.poly }
+
+    /// Borrow the underlying boundary polynomial.
+    pub fn polynomial_ref(&self) -> &DensePolynomial<F> { &self.poly }
+
+    /// Return the evaluations over D of the boundary polynomial, consuming `self`.
+    pub fn evals(self) -> Evaluations<F> { self.evals }
+
+    /// Return the evaluations over D of the boundary polynomial, borrowing `self`.
+    pub fn evals_ref(&self) -> &Evaluations<F> { &self.evals }
+
+    /// Compute the boundary polynomial given a coboundary polynomial
+    /// evaluations `poly_evals` over the elements of a domain D.
+    pub fn from_coboundary_polynomial_evals(
+        poly_evals: Evaluations<F>
+    ) -> Result<Self, Error>
+    {
+        let domain = poly_evals.domain;
+        let evals = poly_evals.evals;
+
+        // Z(1) = 0, or any arbitrary value
+        let mut z_evals = Vec::with_capacity(evals.len());
+        z_evals.push(F::zero());
+
+        // The other coefficients of the boundary polynomial will be the cumulative sum
+        // of the evaluations of the coboundary poly over the domain, e.g.:
+        // Z(g) = Z(1) + p'(g),
+        // ....
+        // Z(g^|H| - 1) = Z(g^|H| - 2) + p'(g^|H| - 1) = p'(g) + p'(g^2) + ... + p'(g^|H - 1|)
+        // Z(g^|H|) = 0 = p'(g) + p'(g^2) + ... + p'(g^|H - 1|) + p'(g^|H|), will be excluded
+        //TODO: Prefix sum here. Parallelize ? Is it worth it ? (rayon (crossbeam too) has no parallel impl for scan())
+        let mut poly_cum_sum_evals = evals.into_iter().scan(F::zero(), |acc, x| {
+            *acc += x;
+            Some(*acc)
+        }).collect::<Vec<_>>();
+
+        // Poly evals over domain should sum to zero
+        if poly_cum_sum_evals[poly_cum_sum_evals.len() - 1] != F::zero() {
+            Err(Error::InvalidPolynomial("The given coboundary polynomial evaluations over the given domain don't sum to zero"))?
+        }
+
+        z_evals.append(&mut poly_cum_sum_evals);
+        z_evals.pop(); // Pop the last zero
+
+        let z_evals = Evaluations::from_vec_and_domain(
+            z_evals,
+            domain,
+        );
+
+        let z_poly = z_evals.interpolate_by_ref();
+
+        Ok(Self {
+            poly: z_poly,
+            evals: z_evals
+        })
+    }
+
+    /// Compute the boundary polynomial given a coboundary
+    /// polynomial `poly` over a domain `domain`.
+    pub fn from_coboundary_polynomial(
+        poly:   &DensePolynomial<F>,
+        domain: Box<dyn EvaluationDomain<F>>
+    ) -> Result<Self, Error>
+    {
+        Self::from_coboundary_polynomial_evals(poly.evaluate_over_domain_by_ref(domain))
+    }
+
+    /// Compute the boundary polynomial given a non-coboundary polynomial
```

Comment:
This comment is a bit misleading. I'd suggest the following:

Given a polynomial p(X) with non-zero domain sum v= Sum_{x in D} p(x), we construct a boundary polynomial Z(X) for the centered polynomial p'(X) = p(X) - v/|D|, and return both Z(X) and v/|D|.